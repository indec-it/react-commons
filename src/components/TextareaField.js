import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {
    Label, Input, FormGroup, FormFeedback
} from 'reactstrap';
import {size} from 'lodash';

import {ValidatorService} from '../services';

class TextareaField extends PureComponent {
    static propTypes = {
        onChange: PropTypes.func.isRequired,
        control: PropTypes.string.isRequired,
        label: PropTypes.string,
        feedBackLabel: PropTypes.string,
        value: PropTypes.string.isRequired,
        minLength: PropTypes.number,
        maxLength: PropTypes.number,
        disabled: PropTypes.bool,
        feedBack: PropTypes.bool
    };

    static defaultProps = {
        label: null,
        maxLength: 250,
        minLength: 2,
        disabled: false,
        feedBack: false,
        feedBackLabel: 'Valor invalido'
    };

    constructor(props) {
        super(props);
        this.state = {valid: undefined};
    }

    validateInput(value, onChange) {
        const {maxLength, minLength} = this.props;
        const valid = ValidatorService.validateText(value, maxLength, minLength);
        this.setState(() => ({valid}), onChange());
    }

    handleChange({target}) {
        if (this.props.value === target.value) {
            return;
        }
        this.validateInput(target.value, this.props.onChange({target}));
    }

    render() {
        const {
            control, label, value, maxLength, minLength, disabled, feedBack, feedBackLabel
        } = this.props;
        const {valid} = this.state;
        return (
            <FormGroup>
                {label && (
                    <Label>
                        {label}
                    </Label>
                )}
                <Input
                    componentClass="textarea"
                    {...{
                        value, disabled, maxLength, minLength, valid
                    }}
                    required
                    onChange={e => this.handleChange(e)}
                    invalid={valid === false}
                    name={control}
                />
                {feedBack && (
                    <FormFeedback tooltip>
                        {feedBackLabel}
                    </FormFeedback>
                )}
                <small>
                    {`Quedan ${maxLength - size(value)} caracter(es) disponibles.`}
                </small>
            </FormGroup>

        );
    }
}

export default TextareaField;
