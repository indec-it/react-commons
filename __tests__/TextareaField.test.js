import React from 'react';
import {shallow} from 'enzyme';

import TextareaField from '../src/components/TextareaField';

const onChange = jest.fn();

describe('<TextareaField/>', () => {
    it('should renders correctly', () => {
        const wrapper = shallow(<TextareaField control="textArea" onChange={onChange} value="text"/>);
        expect(wrapper).toMatchSnapshot();
    });
});
