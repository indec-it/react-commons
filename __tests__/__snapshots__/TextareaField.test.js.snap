// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextareaField/> should renders correctly 1`] = `
<Fragment>
  <TextField
    control="textArea"
    disabled={false}
    feedBackLabel="Valor invalido"
    handleKeyPress={null}
    label={null}
    maxLength={50}
    minLength={2}
    onBlur={null}
    onChange={[MockFunction]}
    placeholder=""
    required={false}
    row={false}
    type="textarea"
    validateInput=""
    value=""
  />
  <small>
    Quedan 246 caracter(es) disponibles.
  </small>
</Fragment>
`;
